src/Widgets/BridgeWidgetApi.ts

```diff
@@ -1,69 +1,211 @@
+            throw Error('Cannot get connections without a valid userId');
+        }
+        await assertUserPermissionsInRoom(req.userId, req.params.roomId as string, "read", this.intent);
+        const allConnections = this.connMan.getAllConnectionsForRoom(req.params.roomId as string);
+        const powerlevel = new PowerLevelsEvent({content: await this.intent.underlyingClient.getRoomStateEvent(req.params.roomId, "m.room.power_levels", "")});
+        const connections = allConnections.map(c => c.getProvisionerDetails?.(true)).filter(c => c?.service === req.params.service).filter(c => !!c) as GetConnectionsResponseItem[];
+        const userPl = powerlevel.content.users?.[req.userId] || powerlevel.defaultUserLevel;
 
-    private async getVerifyToken(req: Request, res: Response) {
-        const roomOrError = await this.getRoomFromRequest(req);
-        if (roomOrError instanceof AdminRoom) {
-            return res.sendStatus(204);
+        for (const c of connections) {
+            const requiredPl = Math.max(powerlevel.content.events?.[c.type] || 0, powerlevel.defaultStateEventLevel);
+            c.canEdit = userPl >= requiredPl;
+            if (!c.canEdit) {
+                delete c.secrets;
+            }
         }
 
-        return res.status(roomOrError.statusCode).send({error: roomOrError.error});
+        res.send({
+            connections,
+            canEdit: userPl >= powerlevel.defaultUserLevel
+        });
     }
 
-    private async getRoomState(req: Request, res: Response) {
-        const roomOrError = await this.getRoomFromRequest(req);
-        if (!(roomOrError instanceof AdminRoom)) {
-            return res.status(roomOrError.statusCode).send({error: roomOrError.error});
+    private async createConnection(req: ProvisioningRequest, res: Response<GetConnectionsResponseItem>) {
+        if (!req.userId) {
+            throw Error('Cannot get connections without a valid userId');
         }
+        await assertUserPermissionsInRoom(req.userId, req.params.roomId as string, "write", this.intent);
         try {
-            return res.send(await roomOrError.getBridgeState());
+            if (!req.body || typeof req.body !== "object") {
+                throw new ApiError("A JSON body must be provided", ErrCode.BadValue);
+            }
+            const connection = await this.connMan.provisionConnection(req.params.roomId as string, req.userId, req.params.type as string, req.body as Record<string, unknown>);
+            if (!connection.getProvisionerDetails) {
+                throw new Error('Connection supported provisioning but not getProvisionerDetails');
+            }
+            res.send(connection.getProvisionerDetails(true));
         } catch (ex) {
-            log.error(`Failed to get room state:`, ex);
-            return res.status(500).send({error: "An error occured when getting room state"});
+            log.warn(`Failed to create connection for ${req.params.roomId}`, ex);
```

Comment:
Why was this degraded to a warning? I guess that makes sense for ApiErrors (maybe even less than warn in that case), but for regular errors this deserves a log.error imo.

(Also, doesn't the addRoute wrapper add its own logging whenever it catches something?)

Reply:
I suppose it could be an error yeah!

The addRoute wrapper logs some stuff, but I wanted some extra detail about what's failing so that's why we catch and rethrow.